src/data/finsum.lean

```diff
@@ -0,0 +1,749 @@
+
+variable {α : Type u}
+
+/-- `univ' α` is `finset.univ` if `α` is a fintype. -/
+@[simp] lemma univ'_eq_univ [h : fintype α] : univ' α = univ :=
+by convert (dif_pos (nonempty.intro h))
+
+lemma univ'_eq_empty (h : ¬ nonempty (fintype α)) : univ' α = ∅ :=
+dif_neg h
+
+lemma univ'_eq_empty' (h : fintype α → false) : univ' α = ∅ :=
+univ'_eq_empty $ not_nonempty_fintype.2 $ ⟨h⟩
+
+lemma mem_univ' [h : fintype α] (x : α) : x ∈ univ' α :=
+(@univ'_eq_univ _ h).symm ▸ mem_univ _
+
+@[simp] lemma mem_univ'' [h : nonempty (fintype α)] (x : α) : x ∈ univ' α :=
+@mem_univ' α (nonempty.some h) x
+
+end finset
+
+/-! # noncomputable finite sums -/
+
+section finsum
+
+variables {α : Type u} {M : Type v} [add_comm_monoid M]
+
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+  otherwise. -/
+noncomputable def finsum (f : α → M) : M :=
+(function.support f).to_finset'.sum f
+
+/-- Sum of `f x` for `x ∈ s`, if `s ∩ f.support` is finite. Zero if it's infinite.  -/
+noncomputable def finsum_in (s : set α) (f : α → M) : M :=
+finsum (λ x, if x ∈ s then f x else 0)
+
+localized "notation `∑ᶠ` binders `, ` r:(scoped:67 f, finsum f) := r"
+  in big_operators
+
+localized "notation `∑ᶠ` binders ` in ` s `, ` r:(scoped:67 f, finsum_in s f) := r"
+  in big_operators
+
+open_locale big_operators
+
+lemma finsum_def (f : α → M) :
+  ∑ᶠ i : α, f i = if h : (function.support f).finite then finset.sum (h.to_finset) f else 0 :=
+begin
+  split_ifs,
+  { unfold finsum set.to_finset',
+    rw dif_pos h },
+  { unfold finsum,
+    convert finset.sum_empty,
+    exact dif_neg h }
+end
+
+lemma finsum_in_def (s : set α) (f : α → M) :
+  ∑ᶠ i in s, f i = finsum (λ x, if x ∈ s then f x else 0) := rfl
+
+-- function.support doesn't import data.finsupp and vice versa
+-- so the below lemma can't go in either place
+lemma function.support_eq_support {α : Type u} {β : Type v} [has_zero β] (f : α →₀ β) :
+  function.support f = f.support :=
+set.ext $ λ x, finsupp.mem_support_iff.symm
+
+namespace finsupp
+
+variables {β : Type v} [has_zero β]
+
+lemma support_finite (f : α →₀ β) : (function.support f).finite :=
+by { rw function.support_eq_support, exact f.support.finite_to_set }
+
+/-- The natural `finsupp` induced by the function `f` given that it has finite support. -/
+noncomputable def of_support_finite
+  (f : α → β) (hf : (function.support f).finite) : α →₀ β :=
+{ support := hf.to_finset,
+  to_fun := f,
+  mem_support_to_fun := λ _, hf.mem_to_finset }
+
+lemma of_support_finite_def {f : α → β} {hf : (function.support f).finite} :
+  (of_support_finite f hf : α → β) = f := rfl
+
+end finsupp
+
+lemma finsum_eq_finsupp_sum (f : α →₀ M) : ∑ᶠ i : α, f i = finsupp.sum f (λ x m, m) :=
+begin
+  rw [finsum_def, dif_pos (finsupp.support_finite f)],
+  congr',
+  simp
+end
+
+lemma finsum_eq_finsupp_sum' (f : α → M) (hf : (function.support f).finite) :
+  ∑ᶠ i : α, f i = finsupp.sum (finsupp.of_support_finite f hf) (λ x m, m) :=
```

Comment:
A more general fact: if `s` is a `finset` such that `∀ x, f x ≠ 0 → x ∈ s`, then `finsum f = s.sum f`.

Reply:
Now added `finsum_eq_finsupp_sum_of_ne_zero_mem_finset`